### DISCUSSION TITLE
Support for  TLS using local CAs with EC keys

### DISCUSSION THREAD

**Discussion by aweatherguy** [arendst/Tasmota#15448] [Ideas]
This is an option that might save a significant amount of flash memory. I could use a little help exploring this possibility, and if anyone is interested let me know. Specifically, from someone who's fluent in BearSSL.

It's possible today to configure TLS with a locally generated CA certificate with (2048-bit) RSA keys in an isolated local network. Devices running Tasmota use certs with 256-bit EC keys which are considerably smaller. For best security the CA cert s/b embedded in the device as opposed to generating a fingerprint upon first connection with a broker, and this requires not only space in flash for the CA cert, but also code in the device to implement RSA algorithms. 

Using a local CA certificate with _EC keys_ (instead of RSA keys) may save significant space in flash for two reasons. First, the CA cert will be smaller and embedding it will require less space. Secondly, there would be no need to instantiate RSA algorithms in the binary image and that might save space, assuming that additional required EC algorithm implementations take less space.

I have experimented with this a bit, and configured BearSSL as best I can. When the device attempts to make a TLS connection, the initial Client Hello and Server Hello response look okay, but then the device appears to crash (Web console resets). 

I have not done further work to troubleshoot this (e.g. increasing logging levels and looking at serial output), but thought I would post here first to see if anyone has the desire and time to offer suggestions...

**Comment by sfromis** [comment]
Would this require more or less CPU time for processing? ESP8266 is already barely fast enough.

**Comment by s-hadinger** [comment]
From my early tests with ECDSA and EC certificates, it was significantly slower than RSA. This is why I dropped it despite the certificate being much smaller

**Comment by aweatherguy** [comment]
Thanks very much for the info...I would have spent many hours exploring a rabbit hole otherwise!
